numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -329,6 +335,98 @@ array_implement_array_function(
 }
 
 
+/*
+ * Implements the __array_function__ protocol for a Python function, as described in
+ * in NEP-18. See numpy.core.overrides for a full docstring.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_array_function(
+    PyObject *NPY_UNUSED(dummy), PyObject *positional_args)
+{
+    PyObject *implementation, *public_api, *relevant_args, *args, *kwargs;
+
+    if (!PyArg_UnpackTuple(
+            positional_args, "implement_array_function", 5, 5,
+            &implementation, &public_api, &relevant_args, &args, &kwargs)) {
+        return NULL;
+    }
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries.
+     */
+    if (kwargs != NULL && PyDict_Contains(kwargs, npy_ma_str_like)) {
+        PyDict_DelItem(kwargs, npy_ma_str_like);
+    }
+
+    PyObject *res = array_implement_array_function_internal(
+        public_api, relevant_args, args, kwargs);
+
+    if (res == Py_NotImplemented) {
+        return PyObject_Call(implementation, args, kwargs);
+    }
+    return res;
+}
+
+
+/*
+ * Implements the __array_function__ protocol for C array creation functions
+ * only. Added as an extension to NEP-18 in an effort to bring NEP-35 to
+ * life with minimal dispatch overhead.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_c_array_function_creation(
+    const char *function_name, PyObject *args, PyObject *kwargs)
+{
+    PyObject *relevant_args;
+    PyObject *numpy_module = NULL, *public_api = NULL;
+    PyObject *result = NULL;
+
+    if (kwargs == NULL) {
+        return Py_NotImplemented;
+    }
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries. If that key isn't present, return NULL and
+     * let originating call to continue.
+     */
+    if (PyDict_Contains(kwargs, npy_ma_str_like)) {
+        relevant_args = PyTuple_Pack(1,
+                PyDict_GetItem(kwargs, npy_ma_str_like));
+        if (relevant_args == NULL) {
+            return NULL;
+        }
+        PyDict_DelItem(kwargs, npy_ma_str_like);
+    }
+    else {
+        return Py_NotImplemented;
+    }
+
+    numpy_module = PyImport_Import(npy_ma_str_numpy);
```

Comment:
```suggestion
    PyObject *numpy_module = PyImport_Import(npy_ma_str_numpy);
```
Then you can remove the initialization to null above.

Reply:
Done in https://github.com/numpy/numpy/pull/16935/commits/91c5ebad90c6d80841d6627cdb8bbce12ba65ce8 . I kept the declaration at the top, as per the standard used throughout the rest of the file.

Reply:
> as per the standard used throughout the rest of the file.

This standard is by accident, as we only recently upgraded to a compiler which allows the other style. I would recommend switching to the C99 style.

Reply:
C99 is specified in the [C style guide](https://numpy.org/neps/nep-0045-c_style_guide.html).

Reply:
Done in https://github.com/numpy/numpy/pull/16935/commits/5202f83e3121c05c1ba619bd6a2934541242a034 , I reformatted the whole file where possible.

Reply:
Normally best to reformat only the changed lines / functions.